proxy/agent.go

```diff
@@ -159,23 +178,22 @@ func (a *agent) Run(stop <-chan struct{}) {
 
 			// delete epoch record and update current config
 			delete(a.epochs, status.epoch)
-			a.current = a.epochs[a.latestEpoch()]
+			a.currentConfig = a.epochs[a.latestEpoch()]
 
 			// schedule a retry for a transient error
-			if status.err != nil && !reflect.DeepEqual(a.desired, a.current) {
-				if a.retryBudget > 0 {
-					delay = a.retryInterval * (2 << uint(a.maxRetries-a.retryBudget))
-					a.retryBudget = a.retryBudget - 1
-					glog.V(2).Infof("Schedule retry after %v (budget %d)", delay, a.retryBudget)
+			if status.err != nil && !reflect.DeepEqual(a.desiredConfig, a.currentConfig) {
+				if a.retry.budget > 0 {
+					a.retry.delay = a.retry.initialInterval * (1 << uint(a.retry.maxRetries-a.retry.budget))
+					a.retry.budget = a.retry.budget - 1
 				} else {
 					glog.Warningf("Permanent error: budget exhausted trying to fulfill the desired configuration")
 					// TODO: update proxy agent monitoring status about this error
-					delay = defaultDelay
+					a.retry.delay = a.retry.defaultDelay
 				}
 			}
 
-		case <-time.After(delay):
-			glog.V(2).Infof("Reconciling after delay %v", delay)
+		case <-time.After(a.retry.delay):
+			glog.V(2).Infof("Reconciling after delay %v (budget %d)", a.retry.delay, a.retry.budget)
```

Comment:
Shouldn't retry.budget be decremented here as well?
Best is to move the logic that changes the values in retry inside reconcile(), so they get adjusted after each reconcile().